src/H5Shyper.c

```diff
@@ -3215,11 +3217,11 @@ H5S__hyper_copy(H5S_t *dst, const H5S_t *src, bool share_selection)
     assert(dst);
 
     /* Allocate space for the hyperslab selection information */
-    if (NULL == (dst->select.sel_info.hslab = H5FL_MALLOC(H5S_hyper_sel_t)))
```

Comment:
The code here and below was directly manipulating the destination dataspace, which could leave things in a bad state if failure occurs. This function had to be reworked to use the local `dst_hslab` and only manipulate the destination dataspace after things have succeeded.